src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,337 @@
+///    [`a.dealloc(ptr, k);`][`dealloc`], and
+///
+///  * if an allocator does not support overallocating, it is fine to simply return
+///    [`layout.size()`] as the actual size.
+///
+/// # Safety
+///
+/// * Pointers returned from an allocator must point to valid memory and retain their validity until
+///   the instance and all of its clones are dropped,
+///
+/// * cloning or moving the allocator must not invalidate pointers returned from this allocator.
+///   A cloned allocator must behave like the same allocator, and
+///
+/// * any pointer to a memory block which is [*currently allocated*] may be passed to any other
+///   method of the allocator.
+///
+/// [*currently allocated*]: #currently-allocated-memory
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub unsafe trait AllocRef {
+    /// On success, returns a pointer meeting the size and alignment guarantees of `layout` and the
+    /// actual size of the allocated block, which is greater than or equal to `layout.size()`.
+    ///
+    /// If this method returns an `Ok((addr, actual_size))`, then the `addr` returned will be
+    /// non-null address pointing to a block of storage suitable for holding an instance represented
+    /// by `layout`.
+    ///
+    /// The returned block of storage is initialized like specified by [`init`], all the way up to
+    /// the returned `actual_size`.
+    ///
+    /// [`init`]: AllocInit
+    ///
+    /// # Errors
+    ///
+    /// Returning `Err` indicates that either memory is exhausted or `layout` does not meet
+    /// allocator's size or alignment constraints.
+    ///
+    /// Implementations are encouraged to return `Err` on memory exhaustion rather than panicking or
+    /// aborting, but this is not a strict requirement. (Specifically: it is *legal* to implement
+    /// this trait atop an underlying native allocation library that aborts on memory exhaustion.)
+    ///
+    /// Clients wishing to abort computation in response to an allocation error are encouraged to
+    /// call the [`handle_alloc_error`] function, rather than directly invoking `panic!` or similar.
+    ///
+    /// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+    fn alloc(&mut self, layout: Layout, init: AllocInit) -> Result<(NonNull<u8>, usize), AllocErr>;
+
+    /// Deallocate the memory referenced by `ptr`.
+    ///
+    /// # Safety
+    ///
+    /// * `ptr` must denote a block of memory [*currently allocated*] via this allocator,
+    ///
+    /// * `layout` must [*fit*] that block of memory, and
+    ///
+    /// * the alignment of the `layout` must match the alignment used to allocate that block of
+    ///   memory.
+    ///
+    /// [*currently allocated*]: #currently-allocated-memory
+    /// [*fit*]: #memory-fitting
+    unsafe fn dealloc(&mut self, ptr: NonNull<u8>, layout: Layout);
+
+    /// Returns a pointer and the actual size of the allocated block. The pointer is suitable for
```

Comment:
The first sentence of the doc comment should explain what the method actually *does*. For example: "Attempts to extend the allocation referenced by `ptr` to fit `new_size`."